File: codes/go/gemini/complete/task_1/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"sort"
)

func allocateBed(patient map[string]interface{}, wards []interface{}) map[string]interface{} {
	patientID := patient["id"].(string)
	patientGender := patient["gender"].(string)
	patientAge := int(patient["age"].(float64))
	patientCondition := patient["condition"].(string)
	patientSpecialRequirements, ok := patient["special_requirements"].([]interface{})
	if !ok {
		patientSpecialRequirements = []interface{}{}
	}

	result := map[string]interface{}{
		"patient_id":                 patientID,
		"assigned_ward":              nil,
		"assigned_bed":               nil,
		"assigned_satellite_hospital": nil,
	}

	for _, wardInterface := range wards {
		ward := wardInterface.(map[string]interface{})
		wardName := ward["ward_name"].(string)
		genderRestriction := ward["gender_restriction"].(string)
		ageRestriction := ward["age_restriction"].([]interface{})
		minAge := int(ageRestriction[0].(float64))
		maxAge := int(ageRestriction[1].(float64))
		targetCondition := ward["target_condition"].(string)
		wardSpecialRequirements, ok := ward["special_requirements"].([]interface{})
		if !ok {
			wardSpecialRequirements = []interface{}{}
		}
		availableBedsInterface, ok := ward["available_beds"].([]interface{})
		if !ok {
			availableBedsInterface = []interface{}{}
		}
		overflowCapacity := int(ward["overflow_capacity"].(float64))
		satelliteHospitalsInterface, ok := ward["satellite_hospitals"].([]interface{})
		if !ok {
			satelliteHospitalsInterface = []interface{}{}
		}

		// Ward Matching
		conditionMatch := patientCondition == targetCondition
		genderMatch := false
		if genderRestriction == "No Restriction" || genderRestriction == "No Gender Restriction" {
			genderMatch = true
		} else if genderRestriction == "Male Only" && patientGender == "Male" {
			genderMatch = true
		} else if genderRestriction == "Female Only" && patientGender == "Female" {
			genderMatch = true
		}
		ageMatch := patientAge >= minAge && patientAge <= maxAge
		specialRequirementsMatch := true
		patientReqMap := make(map[interface{}]bool)
		for _, req := range patientSpecialRequirements {
			patientReqMap[req] = true
		}
		for _, req := range patientSpecialRequirements {
			found := false
			for _, wardReq := range wardSpecialRequirements {
				if req == wardReq {
					found = true
					break
				}
			}
			if !found {
				specialRequirementsMatch = false
				break
			}
		}

		if conditionMatch && genderMatch && ageMatch && specialRequirementsMatch {
			// Regular Bed Allocation
			availableBeds := make([]int, 0)
			for _, bedInterface := range availableBedsInterface {
				availableBeds = append(availableBeds, int(bedInterface.(float64)))
			}

			if len(availableBeds) > 0 {
				sort.Ints(availableBeds)
				assignedBed := availableBeds[0]
				result["assigned_ward"] = wardName
				result["assigned_bed"] = assignedBed
				return result
			}

			// Overflow Bed Allocation
			if overflowCapacity > 0 {
				result["assigned_ward"] = wardName
				result["assigned_bed"] = fmt.Sprintf("Overflow-1")
				return result
			}

			// Satellite Hospital Allocation
			if len(satelliteHospitalsInterface) > 0 {
				satelliteHospitals := make([]string, 0)
				for _, hospitalInterface := range satelliteHospitalsInterface {
					satelliteHospitals = append(satelliteHospitals, hospitalInterface.(string))
				}
				result["assigned_satellite_hospital"] = satelliteHospitals[0]
				result["assigned_ward"] = nil
				result["assigned_bed"] = nil
				return result
			}
		}
	}

	return result
}

func main() {
	var input map[string]interface{}
	decoder := json.NewDecoder(SystemIn{})
	err := decoder.Decode(&input)
	if err != nil {
		panic(err)
	}

	patient := input["patient"].(map[string]interface{})
	wards := input["wards"].([]interface{})

	result := allocateBed(patient, wards)

	jsonResult, err := json.Marshal(result)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(jsonResult))
}

type SystemIn struct{}

func (s SystemIn) Read(p []byte) (n int, err error) {
	return fmt.Scan(string(p))
}

File: codes/python/gemini/complete/task_1/Solution.py
def allocate_bed(patient, wards):
    patient_id = patient['id']
    assigned_ward = None
    assigned_bed = None
    assigned_satellite_hospital = None

    for ward in wards:
        # Ward Matching
        if ward['target_condition'] != patient['condition']:
            continue

        gender_restriction = ward['gender_restriction']
        if gender_restriction == 'Male Only' and patient['gender'] != 'Male':
            continue
        if gender_restriction == 'Female Only' and patient['gender'] != 'Female':
            continue
        if gender_restriction not in ('No Restriction', 'No Gender Restriction') and \
           gender_restriction != patient['gender']:
            continue

        age_restriction = ward['age_restriction']
        if not (age_restriction[0] <= patient['age'] <= age_restriction[1]):
            continue

        ward_special_requirements = ward.get('special_requirements') or []
        patient_special_requirements = patient.get('special_requirements') or []
        if not all(req in ward_special_requirements for req in patient_special_requirements):
            continue

        # Bed Allocation
        if ward['available_beds']:
            assigned_bed = min(ward['available_beds'])
            ward['available_beds'].remove(assigned_bed)
            assigned_ward = ward['ward_name']
            break
        elif ward['overflow_capacity'] > 0:
            assigned_ward = ward['ward_name']
            overflow_beds_used = 0
            assigned_bed = f"Overflow-{overflow_beds_used + 1}"
            ward['overflow_capacity'] -= 1
            break
        elif ward['satellite_hospitals']:
            assigned_satellite_hospital = ward['satellite_hospitals'][0]
            assigned_ward = None
            assigned_bed = None
            break

    return {
        'patient_id': patient_id,
        'assigned_ward': assigned_ward,
        'assigned_bed': assigned_bed,
        'assigned_satellite_hospital': assigned_satellite_hospital
    }


if __name__ == '__main__':
    import json
    import sys

    input_data = json.load(sys.stdin)
    patient = input_data['patient']
    wards = input_data['wards']

    result = allocate_bed(patient, wards)
    print(json.dumps(result))